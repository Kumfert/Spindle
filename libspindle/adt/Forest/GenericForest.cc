//
// GenericForest.cc
//
//  $Id: GenericForest.cc,v 1.2 2000/02/18 01:31:44 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//

#include "spindle/GenericForest.h"

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

#ifdef __FUNC__
#undef __FUNC__
#endif

SPINDLE_IMPLEMENT_PERSISTANT( GenericForest, SpindlePersistant )

GenericForest::GenericForest() {
  // create an empty class
  incrementInstanceCount( GenericForest::MetaData );
  reset();
}

GenericForest::GenericForest(const int n) {
  // creates a Forest of n trees
  incrementInstanceCount( GenericForest::MetaData );
  nNodes = n;
  nTrees = 0;
  firstRoot = -1;
  parent.resize( nNodes );       
  parent.init( -1 );
  child.resize( nNodes );        
  child.init( -1 );
  nextSibling.resize( nNodes );  
  nextSibling.init(  1, 1 ); // counts from 1 to size, step=+1
  prevSibling.resize( nNodes );  
  prevSibling.init( -1, 1 ); // counts from -1 to size-2, step=+1
  nChildren.resize( nNodes );    
  nChildren.init( 0 );
  
  // final touches
  nextSibling[nNodes-1] = -1;
  prevSibling[0] = -1;
  nTrees = nNodes;
  firstRoot = 0;
  currentState = VALID;
}

GenericForest::GenericForest( const int n, int* Parent ) {
  // create a class based on a parent pointer
  incrementInstanceCount( GenericForest::MetaData );
  // non-const so use this parent pointer.
  nNodes = 0;
  nTrees = 0;
  firstRoot = -1;
  resize(n);
  parent.take( Parent, n );
  currentState = UNKNOWN;
  createChildSibling();
  createDoubleLink();
}


GenericForest::GenericForest( const int n, const int* Parent ) { 
  incrementInstanceCount( GenericForest::MetaData );
  // non-const so use this parent pointer.
  nNodes = 0;
  nTrees = 0;
  firstRoot = -1;
  resize(n);
  parent.import( Parent, n );

  currentState = UNKNOWN;
  createChildSibling();
  createDoubleLink();
}

GenericForest::GenericForest(const int n, int* Parent, int *Child, int* Sibling) { 
  incrementInstanceCount( GenericForest::MetaData );
  // non-const structures, so we use the given copy
  nNodes = n;
  nTrees = 0;
  firstRoot = -1;
  parent.take( Parent, n );
  child.take( Child, n );
  nextSibling.take( Sibling, n );

  currentState = UNKNOWN;
  createDoubleLink();
}

GenericForest::GenericForest(const int n, const int* Parent, const int *Child, const int* Sibling) { 
  incrementInstanceCount( GenericForest::MetaData );
  // const structures, so we make a private copy
  nNodes = n;
  nTrees = 0;
  firstRoot = -1;

  // first allocate our own space.
  parent.import(Parent, nNodes);
  child.import(Child, nNodes);
  nextSibling.import(Sibling, nNodes);

  currentState = UNKNOWN;
  createDoubleLink();
}

GenericForest::~GenericForest() {
  decrementInstanceCount( GenericForest::MetaData );
}

bool
GenericForest::reset() {
  nNodes = 0;
  nTrees = 0;
  firstRoot = -1;
  currentState = EMPTY;
  return true;
}

void
GenericForest::resize( const int n ) {
  nNodes = n;
  parent.resize( nNodes );       
  child.resize( nNodes );        
  nextSibling.resize( nNodes );  
  prevSibling.resize( nNodes );  
  nChildren.resize( nNodes );    
  currentState = EMPTY;
}

bool 
GenericForest::setParent( const int length, const int* Parent ) {
  if ( currentState != EMPTY ) { return false; }
  if ( length <= 0 ) { return false; }
  resize(length);
  parent.import( Parent, length );
  createChildSibling();
  createDoubleLink();
  currentState = UNKNOWN;
  return true;
}


bool 
GenericForest::setParent( const int length, int* Parent ) {
  if ( currentState != EMPTY ) { return false; }
  if ( length <= 0 ) { return false; }
  resize(length);
  parent.take( Parent, length );
  createChildSibling();
  createDoubleLink();
  currentState = UNKNOWN;
  return true;
}


void
GenericForest::validate() {
  if ( currentState != UNKNOWN )  { return ; }
  // should do some checking here
  currentState = VALID;
}

void 
GenericForest::createChildSibling() {
  //
  // PRECONDITIONS: assumes parent and nNodes is valid, nothing else
  //                currentState must equal "UNKNOWN";
  // POSTCONDITION: child and nextSibling are initialized.
  //
  if ( currentState != UNKNOWN ) { return ; }

  // 1 resize and initialize child and nextSibling arrays
  child.resize( nNodes );
  child.init( -1 );

  nextSibling.resize( nNodes );
  nextSibling.init( -1 );

  // 2. traverse the parent pointer in reverse
  //    using nextSibling as a stack and child as the pointer into the stack
  for( int i=nNodes-1; i>=0; --i ) {      // for all nodes (in reverse)
    if ( parent[i] == i ) {
      parent[i] = -1;                     // set parent[root] to -1
    } else if ( parent[i] >= 0 ) {        // else if it has a parent
      nextSibling[i] = child[parent[i]];  //     next sibling is the 1st child of parent       
      child[parent[i]] = i;               //     first child of parent becomes this node.
    }
  } 
}

void 
GenericForest::createDoubleLink() { 
  //
  // PRECONDITION:   currentState == UNKNOWN and nNodes, parent, child, and nextSibling are initialized
  //
  // POSTCONDITION:  nChildren and prevSibling is initialized.  List of roots is initialized
  // 
  nTrees = 0;
  firstRoot = -1;
  {for(int i=0; i<nNodes; i++) { 
    if (parent[i] == i) { parent[i] = -1; } // if representing root by self edge
    if (parent[i] == -1) { 
      nTrees++;
      nextSibling[i] = firstRoot;
      firstRoot = i;
    }
  }}
  
  // next initialize prevSibling array
  // start by making it all -1;
  prevSibling.resize( nNodes );
  prevSibling.init( -1 );

  // next, invert nextSibling array
  {for( int i=0; i<nNodes; i++ ) { 
    if ( nextSibling[i] != -1 ) {
      prevSibling[ nextSibling[i] ] = i;
    }
  }}

  // finally initialize nChildren array
  nChildren.resize( nNodes );
  nChildren.init( 0 );
  {for (int i=0; i<nNodes; i++ ) { 
    node parent_i = parent[i];
    if ( parent_i != -1 ) { 
      nChildren[ parent_i ]++;
    }
  }}
}

void
GenericForest::dump( FILE * stream ) const {
  if ( isValid() ) {
    for( int i=0; i<nNodes; ++i ) {
      fprintf(stream, "%d %d\n", i, parent[i] );
    }
  } else { 
    fprintf(stream, "invalid\n" );
  }
  return;
}
  
bool 
GenericForest::pruneNewTree( const node newTreeRoot ) { 
  if ( currentState != VALID ) { return false; }
  if ( (newTreeRoot < 0) || (newTreeRoot >= nNodes) ) { return false; }
  if ( isRoot( newTreeRoot ) ) { return false; }  // no pruning necessary
  node oldparent = parent[ newTreeRoot ];

  // First remove new root from current sibling list
  node old_sibling = child[ oldparent ];
  if ( old_sibling == newTreeRoot ) {                   // if new root is the first child of old parent
    child[ oldparent ] = nextSibling[ newTreeRoot ]; //   set child of old parent to be next sibling
  }
  node next = nextSibling [ newTreeRoot ];
  node prev = prevSibling [ newTreeRoot ];
  if ( next != -1 ) prevSibling[ next ] = prev;
  if ( prev != -1 ) nextSibling[ prev ] = next;
  nChildren[ oldparent ]--; // reduce number of children of old parent

  // now attach newTreeRoot to the list of roots
  nextSibling[ newTreeRoot ] = firstRoot;
  prevSibling[ firstRoot ] = newTreeRoot;
  firstRoot = newTreeRoot;
  parent[ newTreeRoot ] = -1;
  nTrees++;
  return true;
}

bool GenericForest::graftTreeToOther( const node oldRoot, const node newParent ) { 
  if ( currentState != VALID ) { return false; }
  if ( (oldRoot < 0) || (oldRoot >= nNodes) || (newParent < 0) || (newParent >= nNodes) ) { 
    // bad argument: out of range
    return false;
  }
  if ( oldRoot == newParent ) { return false; } // cannot set parent to itself
  if ( ! isRoot(oldRoot) ) { return false; }  // bad argument: not a root
  if ( nTrees == 1 ) { return false; } // cannot graft when there is only one tree
  // first remove oldRoot from its list of roots.
  if ( firstRoot == oldRoot ) { 
    firstRoot = nextSibling[oldRoot];
  }
  node next = nextSibling [ oldRoot ];
  node prev = prevSibling [ oldRoot ];
  if ( next != -1 ) prevSibling[ next ] = prev;
  if ( prev != -1 ) nextSibling[ prev ] = next;
  nTrees--;

  // Now attach oldRoot to child list of newParent
  nextSibling[ oldRoot ] = child[ newParent ];
  prevSibling[ oldRoot ] = -1;
  child[ newParent ] = oldRoot;
  nChildren[ newParent ]++;
  parent[ oldRoot ] = newParent;
  return true;
}

void 
GenericForest::storeObject( SpindleArchive& ar ) const {
  switch(currentState) {
  case INVALID:
    ar << -1;
    break;
  case EMPTY:
    ar << 0 ;
    break;
  case UNKNOWN:
    ar << 1 ;
    break;
  case VALID:
    ar << 2;
    break;
  default:
    ar << -1;
    break;
  }
  if ( currentState==UNKNOWN || currentState==VALID ) {
    const int *p = parent.lend();
    size_t my_size = nNodes;
    pair< const int*, size_t> my_pair( p, my_size );
    ar << my_pair;
  }
}

#define __FUNC__ "GenericForest::loadObject( SpindleArchive& ar )"
void
GenericForest::loadObject( SpindleArchive& ar ) {
  int state;
  ar >> state;
  switch(state) {
  case -1:
    currentState = INVALID;
    break;
  case 0:
    currentState = EMPTY;
    break;
  case 1:
    currentState = UNKNOWN;
    break;
  case 2:
    currentState = VALID;
    break;
  default:
    currentState = INVALID;
    break;
  }
  if ( (currentState==UNKNOWN) || (currentState==VALID) ) {
    size_t new_size = ar.peekLength();
    nNodes = (int) new_size;
    parent.resize(nNodes);
    int* p = parent.give();
    pair<int*, size_t> new_pair( p, new_size );
    
    ar >> new_pair;
    WARN_IF( new_pair.second != new_size, "Warning, expected size=%d, but extracted only %d.", 
	     new_pair.second, new_size );
    parent.take( new_pair.first, new_pair.second );
  }
  createChildSibling();
  createDoubleLink();
  if ( currentState==VALID ) { // if it was saved as valid, we need to revalidate here.
    currentState=UNKNOWN;
    validate();
  }
}
#undef __FUNC__

constGenericForestPostorderIterator::constGenericForestPostorderIterator (const GenericForest& forest) 
  : myForest(forest) { 
    curNode = myForest.firstRoot; 
    curDepth = 0;
    GenericForest::node next;
    while( ( next = myForest.child[ curNode ] ) != -1 ) { curNode = next; curDepth++; }
}
 

constGenericForestPostorderIterator& 
constGenericForestPostorderIterator::operator++() {
  if ( ! myForest.isLastSibling( curNode ) ) {         // if we have a sibling
    curNode = myForest.nextSibling[ curNode ];        //    advance to sibling.
    GenericForest::node next;                            //    then advance to lowest descendant
    while( ( next = myForest.child[ curNode ] ) != -1 ) { curNode = next; curDepth++; }
  } else if ( ! myForest.isRoot(curNode) ) {           // else  if we have a  parent
    curNode = myForest.parent[ curNode ];             //    advance to parent
    curDepth--;                                        // 
  } else {    	                                   // else (we are at last root)
    curNode = -1;                                      //    set iterator to end()
  } 
  return *this;
}

constGenericForestPreorderIterator& 
constGenericForestPreorderIterator::operator++() {
  if ( ! myForest.isLeaf(curNode) ) {                  // if not at a leaf
    curNode = myForest.child[ curNode ];              //    advance to child
    curDepth++;
  } else if ( ! myForest.isLastSibling( curNode ) ) {  // else if not last sibling
    curNode = myForest.nextSibling[ curNode ];        //    advance to sibling.
  } else {                                             // else 
    for ( ; curNode != -1; curNode = myForest.parent[ curNode ]) {
      //    for each ancestor
      if ( ! myForest.isLastSibling( curNode ) ){      //        if ancestor has a sibling
	curNode = myForest.nextSibling[ curNode ];    //           advance to this 
	return *this;                                  //           return
      }
      curDepth--;
    }
    // ASSERT:  if we get here, then curNode == -1 which is the flag for end.
  }
  return *this;
}
