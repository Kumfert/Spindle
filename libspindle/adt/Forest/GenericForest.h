//
// GenericForest.h
//
//  $Id: GenericForest.h,v 1.2 2000/02/18 01:31:44 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//

#ifndef SPINDLE_ARRAY_FOREST_H
#define SPINDLE_ARRAY_FOREST_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

#ifdef REQUIRE_OLD_CXX_HEADER_SUFFIX
#include "iterator.h"
#else
#include <iterator>
using namespace std;
#endif

SPINDLE_BEGIN_NAMESPACE // 

//
// Iterator Class Declarations
//

class constGenericForestIterator;
class constGenericForestPostorderIterator;
class constGenericForestPreorderIterator;

class constGenericForestIterator;
class constGenericForestPostorderIterator;
class constGenericForestPreorderIterator;

class GenericForest : public SpindlePersistant { 

  SPINDLE_DECLARE_PERSISTANT( GenericForest )

public:
  friend class constGenericForestIterator;
  friend class constGenericForestPostorderIterator;
  friend class constGenericForestPreorderIterator;

  typedef int Key;
  typedef int node;
  typedef constGenericForestIterator const_generic_iterator;
  typedef constGenericForestPostorderIterator const_postorder_iterator;
  typedef constGenericForestPreorderIterator const_preorder_iterator;
 
  // Constructors/Destructors
  GenericForest();                                // creates empty forest
  GenericForest(const int n);                     // creates forest of n trees
  GenericForest(const int n, int* Parent);        // assumes ownership of parent pointer
  GenericForest(const int n, const int* Parent);  // copies parent pointer
  GenericForest(const int n, int* Parent, int *Child, int* Sibling); // assumes ownership of all
  GenericForest(const int n, const int* parentNode, const int* firstChild, const int* nextSibling);
  virtual ~GenericForest();
  
protected:
  SharedArray<int> parent;          // Parent of each node
  void createChildSibling();      // given parent, creates firstChild and nextSibling
  void createDoubleLink();        // given parent, firstChild, and nextSibling, creates nChildren & prevChild

private:
  int nNodes;            // number of nodes in all trees
  int nTrees;            // number of trees 
  int firstRoot;         // root of first tree in the forest
  
  SharedArray<int> child;           // first child of each node
  SharedArray<int> nextSibling;     // next sibling of each node
  SharedArray<int> prevSibling;     // next sibling of each node
  SharedArray<int> nChildren;       // number of children of a node.

 protected:
  // Simple tests
  bool isLeaf( const node cur ) const;
  bool isRoot( const node cur ) const;
  bool isLastSibling( const node cur ) const;
  bool isFirstSibling ( const node cur ) const;
  bool operator==( const GenericForest& f) const;

 public:
  bool setParent( const int length, const int* Parent );
  bool setParent( const int length, int * Parent );


  virtual void validate();
  virtual bool isValid() const;
  virtual void dump( FILE * stream ) const;

  bool reset();
  void resize( const int n );

  int size() const;
  int queryNTrees() const;
  int queryParent( const node Child ) const;
  bool pruneNewTree( const node newTreeRoot );
  bool graftTreeToOther( const node oldRoot, const node newParent );
  const SharedArray<int>& getParent() const { return parent; }

  // Iterator Access
  const_postorder_iterator begin_postorder() const ;
  const_postorder_iterator end_postorder() const ;
  const_preorder_iterator begin_preorder() const ;
  const_preorder_iterator end_preorder() const ;
  const_generic_iterator create_generic_iterator() const ;

};

class constGenericForestIterator { 
  friend class GenericForest;
private:
  const GenericForest& myForest;
  GenericForest::node curNode;
  GenericForest::node curTree;
  int curDepth;
protected:
  // Constructor/Destructor.  
  // NOTE:  protected constructor restricts construction to derived classes and friends
  constGenericForestIterator( const GenericForest& forest);
  
public:
  bool nextTree(); // point to next tree in forest.  Returns false if no more trees exist.
  bool prevTree(); // point to prev tree in forest.  Returns false if no more trees exist.
  bool parent(); // point to the parent of the current node.  Returns false if current is a root.
  bool nextSibling(); // point to next sibling, return false if current is last sibling or 
                      // if current is root.  Use nextTree() to advance to another tree in the forest.
  bool prevSibling(); // point to next sibling, return false if current is last sibling or 
                      // if current is root.  Use prevTree() to advance to another tree in the forest.
  bool firstChild(); // point to the first child of current, false if current is a leaf.
  int queryNSiblings() const ; // return total of siblings, if root, return number of trees in forest.
  int queryNChildren() const ; // return the total number of children.
  int queryDepth() const;
  
  GenericForest::node operator*() const; // dereference pointer
};

class constGenericForestPostorderIterator : public input_iterator< int, ptrdiff_t > { 
  friend class GenericForest;
private:
  const GenericForest& myForest;
  GenericForest::node curNode;
  int curDepth;
protected:
  // Constructor/Destructor.  
  // NOTE:  protected constructor restricts construction to derived classes and friends
  constGenericForestPostorderIterator(const GenericForest& forest);
  constGenericForestPostorderIterator(const GenericForest& forest, int);
  
public:
  bool operator==(const constGenericForestPostorderIterator& p) const;
  GenericForest::node operator*() const;
  
  constGenericForestPostorderIterator& operator++();
  constGenericForestPostorderIterator operator++(int);
  
  bool isRoot() const;
  bool isLeaf() const;
  int  queryDepth() const;
  
}; // end class constGenericForestPostorderIterator

class constGenericForestPreorderIterator : public input_iterator< int, ptrdiff_t > { 
  friend class GenericForest;
private:
  const GenericForest& myForest;
  GenericForest::node curNode;
  int curDepth;
protected:
  // Constructor/Destructor.  
  // NOTE:  protected constructor restricts construction to derived classes and friends
  constGenericForestPreorderIterator(const GenericForest& forest);
  constGenericForestPreorderIterator(const GenericForest& forest, int);
  
public:
  bool operator==(const constGenericForestPreorderIterator& p) const;
  GenericForest::node operator*() const;
  
  constGenericForestPreorderIterator&  operator++();
  constGenericForestPreorderIterator operator++(int);
  
  bool isRoot() const;
  bool isLeaf() const;
  int  queryDepth() const;
  
}; // end class constGenericForestPreorderIterator

////////////////////////////////////////////////////////////
//
// GenericForest inlined functions
//

inline bool
GenericForest::isValid() const { 
  return (currentState == VALID ); 
}

inline bool 
GenericForest::isLeaf( const node cur ) const {
  return ( child[ cur ] == -1 ) ; 
}

inline bool 
GenericForest::isRoot( const node cur ) const {
  return ( parent[ cur ] == -1 ) ; 
}

inline bool 
GenericForest::isLastSibling( const node cur ) const {
  return  ( nextSibling[ cur ] == -1 ) ;  
}


inline bool 
GenericForest::isFirstSibling ( const node cur ) const {
  return ( prevSibling[ cur ] == -1 ) ; 
}

inline bool 
GenericForest::operator==( const GenericForest& f) const {
 return ( (void *) this == (void *) &f ); 
}

inline int 
GenericForest::size() const { 
  return nNodes; 
}

inline int 
GenericForest::queryNTrees() const { 
  return nTrees; 
}

inline int 
GenericForest::queryParent( const node Child ) const { 
  if ((Child<0) || (Child>=nNodes)) { return -1; }
  return  (parent[Child]==-1) ? Child : parent[Child];
}


inline GenericForest::const_postorder_iterator 
GenericForest::begin_postorder() const { 
  return constGenericForestPostorderIterator( *this ); 
}

inline  GenericForest::const_postorder_iterator 
GenericForest::end_postorder() const { 
  return constGenericForestPostorderIterator(*this, 0); 
}

inline  GenericForest::const_preorder_iterator  
GenericForest::begin_preorder() const { 
  return constGenericForestPreorderIterator( *this ); 
}

inline GenericForest::const_preorder_iterator  
GenericForest::end_preorder() const { 
  return constGenericForestPreorderIterator(*this, 0); 
}

inline GenericForest::const_generic_iterator   
GenericForest::create_generic_iterator() const { 
  return constGenericForestIterator( *this ); 
}

//////////////////////////////////////////////////////////////////////////////////////
//
// constGenericForestIterator methods.
//
inline
constGenericForestIterator::constGenericForestIterator( const GenericForest& forest) : myForest(forest) { 
  curNode = curTree = myForest.firstRoot; 
  curDepth = 0;
}

inline bool 
constGenericForestIterator::nextTree() {  
  // point to next tree in forest.  Returns false if no more trees exist.
  if ( myForest.isLastSibling(curTree) ) { return false; }
  curNode = curTree = myForest.nextSibling[curTree];
  curDepth = 0;
  return true;
}

inline bool 
constGenericForestIterator::prevTree() {  
  // point to prev tree in forest.  Returns false if no more trees exist.
  if ( myForest.isFirstSibling(curTree) ) { return false; }
  curNode = curTree = myForest.prevSibling[curTree];
  curDepth = 0;
  return true;
}

inline bool 
constGenericForestIterator::parent() {  
  // point to the parent of the current node.  Returns false if current is a root.
  if ( myForest.isRoot(curNode) ) { return false; }
  curNode = myForest.parent[curNode];
  curDepth--;
  return true;
}

inline bool 
constGenericForestIterator::nextSibling() { 
  // point to next sibling, return false if current is last sibling or 
  // if current is root.  Use nextTree() to advance to another tree in the forest.
  if ( myForest.isRoot(curNode) ) { return false; }
  if ( myForest.isLastSibling(curNode) ) { return false; }
  curNode = myForest.nextSibling[curNode];
  return true;
}

inline bool 
constGenericForestIterator::prevSibling() { 
  // point to next sibling, return false if current is last sibling or 
  // if current is root.  Use nextTree() to advance to another tree in the forest.
  if ( myForest.isRoot(curNode) ) { return false; }
  if ( myForest.isFirstSibling(curNode) ) { return false; }
  curNode = myForest.prevSibling[curNode];
  return true;
}

inline bool 
constGenericForestIterator::firstChild() { // point to the first child of current, false if current is a leaf.
  if ( myForest.isLeaf(curNode) ) { return false; }
  curNode = myForest.child[curNode];
  curDepth++;
  return true;
}

inline int 
constGenericForestIterator::queryNSiblings() const { 
  // return total of siblings, if root, return number of trees in forest.
  return ( myForest.isRoot( curNode ) ) ? myForest.nTrees : myForest.nChildren[ myForest.parent [ curNode ] ] ; 
}
  
inline int 
constGenericForestIterator::queryNChildren() const { 
  // return the total number of children.
  return myForest.nChildren[ curNode ] ;
}

inline int 
constGenericForestIterator::queryDepth() const { 
  return curDepth; 
}

inline GenericForest::node
constGenericForestIterator::operator*() const { 
  return curNode; 
}

////////////////////////////////////////////////////////////////////////////
// 
// constGenericForestPostorderIterator methods
//
inline 
constGenericForestPostorderIterator::constGenericForestPostorderIterator(const GenericForest& forest, int) 
  : myForest(forest) { 
    curNode = -1; 
}
 
inline bool
constGenericForestPostorderIterator::operator==(const constGenericForestPostorderIterator& p) const { 
  return ( ( myForest == p.myForest) && (curNode == p.curNode) );
}
 
inline GenericForest::node 
constGenericForestPostorderIterator::operator*() const { 
  return curNode; 
}

inline constGenericForestPostorderIterator
constGenericForestPostorderIterator::operator++(int) {
  constGenericForestPostorderIterator tmp = *this;
  ++(*this);
  return tmp;
}

inline bool 
constGenericForestPostorderIterator::isRoot() const {
  return myForest.isRoot(curNode); 
}

inline bool 
constGenericForestPostorderIterator::isLeaf() const { 
  return myForest.isLeaf(curNode); 
}

inline int  
constGenericForestPostorderIterator::queryDepth() const { 
  return curDepth; 
}

////////////////////////////////////////////////////////////////////////////////////////
//
// constGenericForestPreorderIterator
// 
inline 
constGenericForestPreorderIterator::constGenericForestPreorderIterator(const GenericForest& forest) 
  : myForest(forest) { 
    curNode = myForest.firstRoot; 
    curDepth = 0;
}

inline 
constGenericForestPreorderIterator::constGenericForestPreorderIterator(const GenericForest& forest, int) 
  : myForest(forest) { 
    curNode = -1; 
}
  
inline bool 
constGenericForestPreorderIterator::operator==(const constGenericForestPreorderIterator& p) const { 
  return ( (myForest == p.myForest) && (curNode == p.curNode) );
}
  
inline GenericForest::Key 
constGenericForestPreorderIterator::operator*() const { 
  return curNode; 
}

inline constGenericForestPreorderIterator
constGenericForestPreorderIterator::operator++(int) {
  constGenericForestPreorderIterator tmp = *this;
  ++(*this);
  return tmp;
}
  
inline	bool 
constGenericForestPreorderIterator::isRoot() const { 
  return myForest.isRoot(curNode); 
}

inline	bool 
constGenericForestPreorderIterator::isLeaf() const { 
  return myForest.isLeaf(curNode); 
}

inline	int  
constGenericForestPreorderIterator::queryDepth() const { 
  return curDepth; 
}

SPINDLE_END_NAMESPACE


#endif
