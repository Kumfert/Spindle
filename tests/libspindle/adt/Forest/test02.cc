//
// testGenericForest_1b.cc
//
//  $Id: test02.cc,v 1.2 2000/02/18 01:32:05 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is povided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//

#include <iostream.h>
#include "spindle/GenericForest.h"
#include  <iomanip.h>

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

typedef GenericForest Forest;

#define TEST(A)  ( ((A)) ? "Pass" : "Fail" )

void preprint( const Forest& f, ostream& outstr = cout, int space = 4 ) {
  outstr << "------ begin forest (preorder) " << f.queryNTrees() << " tree(s) ----- " << endl;
  for ( Forest::const_preorder_iterator cur = f.begin_preorder(), stop = f.end_preorder(); 
       cur != stop; ++cur ) { 
    outstr << setw( ( cur.queryDepth() + 1) * space ) << "" << *cur << endl;
  }
  outstr << "------  end  forest (preorder) ----- " << endl;
}

void postprint( const Forest& f, ostream& outstr = cout, int space = 4 ) {
  outstr << "------ begin forest (postorder) " << f.queryNTrees() << " tree(s) ----- " << endl;
  for ( Forest::const_postorder_iterator cur = f.begin_postorder(), stop = f.end_postorder(); 
       cur != stop; ++cur ) { 
    outstr << setw( ( cur.queryDepth() + 1 ) * space ) << "" << *cur << endl;
  }
  outstr << "------  end  forest (postorder)----- " << endl;
}
    
/*
                    8
                  /   \
                 0     1 
                 |     | \  \
                 4     7 9 10
             / / | \     | \ 
            5 6 12 13    2  3
            |            |     
            14          11 

*/

const int Parent[] = { 8, 8, 10, 10, 0, 4, 4, 1, 8, 1, 1, 2, 4, 4, 5 };

int main() {
  ios::sync_with_stdio();

  cout << "A tree created straight from a parent-pointer" << endl;

  Forest f(15, Parent);
  f.validate();
  if (! f.isValid() ) {
    cerr << "Error: failed to create a valid forest" << endl;
    exit(-1);
  }
  preprint( f ); 

  cout << "  Prune new tree at 7... " << TEST( f.pruneNewTree(7) ) << endl;
  cout << "  Prune new tree at 5... " << TEST( f.pruneNewTree(5) ) << endl;
  cout << "  Prune new tree at 2... " << TEST( f.pruneNewTree(2) ) << endl;

  preprint(f);

  cout << "  Try to make 9 a subtree of 3... " << TEST( f.graftTreeToOther( 9, 7) )
       << "  (should fail because 9 isn't a root)" << endl;
  cout << "  Try to make 1 a subree of 1 ... " << TEST( f.graftTreeToOther( 1, 1) )
       << "  (should fail because node can't be its own parent)" << endl;

  preprint(f);
  
}
