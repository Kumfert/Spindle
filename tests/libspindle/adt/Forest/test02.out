A tree created straight from a parent-pointer
------ begin forest (preorder) 1 tree(s) ----- 
    8
        0
            4
                5
                    14
                6
                12
                13
        1
            7
            9
            10
                2
                    11
                3
------  end  forest (preorder) ----- 
  Prune new tree at 7... Pass
  Prune new tree at 5... Pass
  Prune new tree at 2... Pass
------ begin forest (preorder) 4 tree(s) ----- 
    2
        11
    5
        14
    7
    8
        0
            4
                6
                12
                13
        1
            9
            10
                3
------  end  forest (preorder) ----- 
  Try to make 9 a subtree of 3... Fail  (should fail because 9 isn't a root)
  Try to make 1 a subree of 1 ... Fail  (should fail because node can't be its own parent)
------ begin forest (preorder) 4 tree(s) ----- 
    2
        11
    5
        14
    7
    8
        0
            4
                6
                12
                13
        1
            9
            10
                3
------  end  forest (preorder) ----- 
